File: backend/api/connections.py
"""Экран «Источники»: подключение, состояние, отключение."""
import json
import secrets

from fastapi import APIRouter, Depends, HTTPException, Request, status
from fastapi.responses import RedirectResponse
from sqlalchemy import select
from sqlalchemy.orm import Session

from backend import config
from backend.db import get_db
from backend.models import SOURCE_KINDS, Connection, User, utcnow
from backend.schemas import AuthUrl, ConnectionOut, TelegramConnect
from backend.security import current_user
from backend.sources import gmail, telegram

router = APIRouter(prefix="/api/connections", tags=["connections"])

OAUTH_STATE_KEY = "gmail_oauth_state"


def _get_or_create(db: Session, user: User, kind: str) -> Connection:
    connection = db.execute(
        select(Connection).where(Connection.user_id == user.id, Connection.kind == kind)
    ).scalar_one_or_none()
    if connection is None:
        connection = Connection(user_id=user.id, kind=kind, state="off", account="", credentials="")
        db.add(connection)
        db.commit()
        db.refresh(connection)
    return connection


@router.get("", response_model=list[ConnectionOut])
def list_connections(
    user: User = Depends(current_user),
    db: Session = Depends(get_db),
) -> list[ConnectionOut]:
    """Оба источника всегда в списке: неподключённый показывается как off."""
    existing = {
        connection.kind: connection
        for connection in db.execute(
            select(Connection).where(Connection.user_id == user.id)
        ).scalars()
    }
    result = []
    for kind in SOURCE_KINDS:
        connection = existing.get(kind)
        result.append(
            ConnectionOut(
                kind=kind,
                state=connection.state if connection else "off",
                account=connection.account if connection else "",
                last_sync_at=connection.last_sync_at if connection else None,
            )
        )
    return result


@router.post("/gmail/start", response_model=AuthUrl)
def gmail_start(request: Request, user: User = Depends(current_user)) -> AuthUrl:
    state = secrets.token_urlsafe(16)
    request.session[OAUTH_STATE_KEY] = state
    try:
        return AuthUrl(auth_url=gmail.auth_url(state))
    except gmail.GmailError as error:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(error))


@router.get("/gmail/callback")
def gmail_callback(
    request: Request,
    code: str | None = None,
    state: str | None = None,
    db: Session = Depends(get_db),
    user: User = Depends(current_user),
) -> RedirectResponse:
    expected = request.session.pop(OAUTH_STATE_KEY, None)
    if not code or not state or state != expected:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail="Авторизация Google не подтверждена"
        )
    try:
        email, refresh_token = gmail.exchange_code(code)
    except gmail.GmailError as error:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=str(error))

    connection = _get_or_create(db, user, "gmail")
    connection.account = email
    connection.credentials = json.dumps({"refresh_token": refresh_token})
    connection.state = "active"
    connection.sync_cursor = None  # новое подключение — импорт за 30 дней
    connection.last_sync_at = None
    db.commit()
    return RedirectResponse(url=f"{config.FRONTEND_ORIGIN}/connections?connected=gmail")


@router.post("/telegram", response_model=ConnectionOut)
def connect_telegram(
    payload: TelegramConnect,
    user: User = Depends(current_user),
    db: Session = Depends(get_db),
) -> ConnectionOut:
    """Токен проверяется через getMe — пользователю показываем имя бота."""
    try:
        bot_name = telegram.verify_token(payload.bot_token.strip())
    except telegram.TelegramError as error:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(error))

    connection = _get_or_create(db, user, "telegram")
    connection.account = bot_name
    connection.credentials = json.dumps({"bot_token": payload.bot_token.strip()})
    connection.state = "active"
    connection.last_sync_at = utcnow()
    db.commit()
    db.refresh(connection)
    return ConnectionOut(
        kind=connection.kind,
        state=connection.state,
        account=connection.account,
        last_sync_at=connection.last_sync_at,
    )


@router.delete("/{kind}", status_code=status.HTTP_204_NO_CONTENT)
def disconnect(
    kind: str,
    user: User = Depends(current_user),
    db: Session = Depends(get_db),
):
    """Отключение обнуляет доступы, но не удаляет сообщения — они остаются в ленте."""
    if kind not in SOURCE_KINDS:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Неизвестный источник")
    connection = db.execute(
        select(Connection).where(Connection.user_id == user.id, Connection.kind == kind)
    ).scalar_one_or_none()
    if connection is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Источник не подключён")
    connection.state = "off"
    connection.credentials = ""
    db.commit()
    return None

File: backend/internal/api/connections.go
package api

import (
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"errors"
	"net/http"
	"strings"

	"personalinbox/internal/store"
	"personalinbox/internal/view"
)

// randomState — одноразовое значение state для OAuth: защита от подмены ответа.
func randomState() string {
	raw := make([]byte, 16)
	if _, err := rand.Read(raw); err != nil {
		return "personal-inbox-state"
	}
	return base64.RawURLEncoding.EncodeToString(raw)
}

// handleListConnections: оба источника всегда в списке, неподключённый
// показывается как off.
func (s *Server) handleListConnections(w http.ResponseWriter, r *http.Request) {
	user, ok := s.currentUser(w, r)
	if !ok {
		return
	}
	existing, err := s.db.ConnectionsOf(user.ID)
	if err != nil {
		fail(w, http.StatusInternalServerError, "Не удалось получить источники")
		return
	}
	byKind := map[string]*store.Connection{}
	for _, connection := range existing {
		byKind[connection.Kind] = connection
	}

	result := make([]view.Connection, 0, len(store.SourceKinds))
	for _, kind := range store.SourceKinds {
		if connection, ok := byKind[kind]; ok {
			result = append(result, view.ConnectionOut(connection))
			continue
		}
		result = append(result, view.Connection{Kind: kind, State: "off"})
	}
	respond(w, http.StatusOK, result)
}

func (s *Server) handleGmailStart(w http.ResponseWriter, r *http.Request) {
	user, ok := s.currentUser(w, r)
	if !ok {
		return
	}
	state := randomState()
	authURL, err := s.gmail.AuthURL(state)
	if err != nil {
		fail(w, http.StatusServiceUnavailable, err.Error())
		return
	}
	s.setSession(w, session{UserID: user.ID, OAuthState: state})
	respond(w, http.StatusOK, map[string]string{"auth_url": authURL})
}

func (s *Server) handleGmailCallback(w http.ResponseWriter, r *http.Request) {
	user, ok := s.currentUser(w, r)
	if !ok {
		return
	}
	expected := s.decodeSession(r).OAuthState
	code := r.URL.Query().Get("code")
	state := r.URL.Query().Get("state")
	// state одноразовый: после проверки его в сессии больше нет.
	s.setSession(w, session{UserID: user.ID})
	if code == "" || state == "" || state != expected {
		fail(w, http.StatusBadRequest, "Авторизация Google не подтверждена")
		return
	}

	email, refreshToken, err := s.gmail.ExchangeCode(code)
	if err != nil {
		fail(w, http.StatusBadGateway, err.Error())
		return
	}
	connection, err := s.db.GetOrCreateConnection(user.ID, "gmail")
	if err != nil {
		fail(w, http.StatusInternalServerError, "Не удалось сохранить подключение")
		return
	}
	credentials, _ := json.Marshal(map[string]string{"refresh_token": refreshToken})
	connection.Account = email
	connection.Credentials = string(credentials)
	connection.State = "active"
	// Новое подключение — импорт за 30 дней.
	connection.SyncCursor = ""
	connection.LastSyncAt = nil
	if err := s.db.SaveConnection(connection); err != nil {
		fail(w, http.StatusInternalServerError, "Не удалось сохранить подключение")
		return
	}
	http.Redirect(w, r, s.cfg.FrontendOrigin+"/connections?connected=gmail", http.StatusFound)
}

// handleConnectTelegram: токен проверяется через getMe — пользователю
// показываем имя бота.
func (s *Server) handleConnectTelegram(w http.ResponseWriter, r *http.Request) {
	user, ok := s.currentUser(w, r)
	if !ok {
		return
	}
	var payload struct {
		BotToken string `json:"bot_token"`
	}
	if !readJSON(w, r, &payload) {
		return
	}
	token := strings.TrimSpace(payload.BotToken)
	if len(token) < 10 {
		fail(w, http.StatusUnprocessableEntity, "Токен бота слишком короткий")
		return
	}

	botName, err := s.telegram.VerifyToken(token)
	if err != nil {
		fail(w, http.StatusBadRequest, err.Error())
		return
	}

	connection, err := s.db.GetOrCreateConnection(user.ID, "telegram")
	if err != nil {
		fail(w, http.StatusInternalServerError, "Не удалось сохранить подключение")
		return
	}
	credentials, _ := json.Marshal(map[string]string{"bot_token": token})
	now := store.UTCNow()
	connection.Account = botName
	connection.Credentials = string(credentials)
	connection.State = "active"
	connection.LastSyncAt = &now
	if err := s.db.SaveConnection(connection); err != nil {
		fail(w, http.StatusInternalServerError, "Не удалось сохранить подключение")
		return
	}
	respond(w, http.StatusOK, view.ConnectionOut(connection))
}

// handleDisconnect обнуляет доступы, но не удаляет сообщения — они остаются
// в ленте.
func (s *Server) handleDisconnect(w http.ResponseWriter, r *http.Request) {
	user, ok := s.currentUser(w, r)
	if !ok {
		return
	}
	kind := r.PathValue("kind")
	if !store.Contains(store.SourceKinds, kind) {
		fail(w, http.StatusNotFound, "Неизвестный источник")
		return
	}
	connection, err := s.db.Connection(user.ID, kind)
	if errors.Is(err, store.ErrNotFound) {
		fail(w, http.StatusNotFound, "Источник не подключён")
		return
	}
	if err != nil {
		fail(w, http.StatusInternalServerError, "Не удалось отключить источник")
		return
	}
	connection.State = "off"
	connection.Credentials = ""
	if err := s.db.SaveConnection(connection); err != nil {
		fail(w, http.StatusInternalServerError, "Не удалось отключить источник")
		return
	}
	w.WriteHeader(http.StatusNoContent)
}
